segmentation.py

```diff
@@ -0,0 +1,575 @@
+model.
+
+Segmentation is carried out in an online fashion by recursively updating a set
+of hypotheses. The hypotheses capture the belief about the current segment,
+e.g. its duration and the linear relationship between inputs and outputs, given
+all the data so far. Each time a new pair of data is received, the hypotheses
+are propagated and re-weighted to reflect this new knowledge.
+
+.. codeauthor:: Gabriel Agamennoni <abriel.agamennoni@mavt.ethz.ch>
+.. codeauthor:: Asher Bender <a.bender@acfr.usyd.edu.au>
+
+"""
+import numpy as np
+from numpy import linalg
+from numpy import random
+from scipy import special
+
+
+class MatrixVariateNormalInvGamma(object):
+    """Matrix-variate normal, inverse-gamma distribution.
+
+    The matrix-variate normal, inverse-gamma distribution is the conjugate
+    prior for a matrix-variate normal distribution. As a result the
+    distribution can be used in Bayesian estimation of the location and scale
+    parameters of the matrix-variate normal distribution.
+
+    """
+
+    def __init__(self, mu, omega, sigma, eta):
+
+        # Get size of data.
+        m, n = np.shape(mu)
+        self.__m, self.__n = m, n
+
+        # Check that the location parameter is a matrix of finite numbers.
+        if not (np.ndim(mu) == 2 and
+                np.shape(mu) == (m, n) and
+                not np.isnan(mu).any() and
+                np.isfinite(mu).all()):
+            msg = 'The location parameter must be a matrix of finite numbers.'
+            raise Exception(msg)
+
+        # Check that the scale parameter is a symmetric, positive-definite
+        # matrix.
+        if not (np.ndim(omega) == 2 and
+                np.shape(omega) == (m, m) and
+                not np.isnan(omega).any() and
+                np.isfinite(omega).all() and
+                np.allclose(np.transpose(omega), omega) and
+                linalg.det(omega) > 0.0):
+            msg = 'The scale parameter must be a symmetric, positive-definite'
+            msg += ' matrix.'
+            raise Exception(msg)
+
+        # Check that the dispersion parameter is a symmetric, positive-definite
+        # matrix.
+        if not (np.ndim(sigma) == 2 and
+                np.shape(sigma) == (n, n) and
+                not np.isnan(sigma).any() and
+                np.isfinite(sigma).all() and
+                np.allclose(np.transpose(sigma), sigma) and
+                linalg.det(sigma) > 0.0):
+            msg = 'The noise parameter must be a symmetric, positive-definite'
+            msg += ' matrix.'
+            raise Exception(msg)
+
+        # Check that the shape parameter is a number greater than one minus the
+        # number of degrees of freedom.
+        if not (np.isscalar(eta) and
+                not np.isnan(eta) and
+                np.isfinite(eta) and eta > n - 1.0):
+            msg = 'The shape parameter must be greater than one minus the'
+            msg += ' degrees of freedom.'
+            raise Exception(msg)
```

Comment:
Unfortunately not!

The try catch approach only works when the code is executed. At this stage they not used (only stored). Any errors in specification will occur later in the code (when the error message is no longer as informative).

The sufficient statisitcs class could be made private. At that point people can use it at their own peril. The code checking can then be moved into the inference engine - it would only be performed once per object (not every time the SS are created).
variable sizes for storage. They are used in operator seldom fails